### PR TITLE
[WIP] initial impl

### PR DESCRIPTION
**Pull request by podviaznikov**
Please see [README](https://github.com/Runnable/loki/pull/1/files#diff-04c6e90faac2675aa89e2176d2eec7d8R1) for more details

Reviewers: 
- [x] @thejsj
- [x] @anandkumarpatel

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+language: node_js
+node_js:
+  - "4"
```
Comment (thejsj) [inline .travis.yml:4]: I would also the most recent version of node to make this future proof.

File: index.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
+module.exports.Docker = require('./lib/docker')
+module.exports.Swarm = require('./lib/swarm')
+// for testing
+module.exports._BaseClient = require('./lib/base')
```
Comment (anandkumarpatel) [inline index.js:4]: can you not just `require('./lib/base')` in your test?

File: lib/base.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+  certs.cert = fs.readFileSync(join(certPath, '/cert.pem'))
+  certs.key = fs.readFileSync(join(certPath, '/key.pem'))
+} catch (e) {
+  throw e
+}
+
+/**
+ * Base Docker/Swarm client with batteries. Can be extended
+ * @param {Class} DriverClass docker driver class
+ * @param {Object} opts - docker options
```
Comment (anandkumarpatel) [inline lib/base.js:43]: I would list out all `opts.opts` that are valid here

File: lib/base.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+ * @class
+ * @extends loki:base
+ * @author Anton Podviaznikov
+ */
+class BaseDockerClient {
+
+  constructor (DriverClass, opts) {
+    if (!DriverClass) {
```
Comment (anandkumarpatel) [inline lib/base.js]: no `joi`? fell like this would be better expresses as a `joi` schema. will make input more strict

File: lib/logger.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * Streams for ponos's bunyan logger.
+ * @private
+ * @return {array} An array of streams for the bunyan logger
+ */
+function _getStreams () {
+  return [
+    {
+      level: process.env.LOG_LEVEL,
```
Comment (anandkumarpatel) [inline lib/logger.js:27]: `LOG_LEVEL_STDOUT`?

File: lib/base.js (hunk 4 above)
Comment (podviaznikov) [inline lib/base.js]: this is a good idea

File: lib/base.js [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+  certs.ca = fs.readFileSync(join(certPath, '/ca.pem'))
+  certs.cert = fs.readFileSync(join(certPath, '/cert.pem'))
+  certs.key = fs.readFileSync(join(certPath, '/key.pem'))
+} catch (e) {
+  throw e
+}
+
+/**
+ * Base Docker/Swarm client with batteries. Can be extended
+ * @param {Class} DriverClass docker driver class
```
Comment (anandkumarpatel) [inline lib/base.js]: name does not match the class defined here ... might be better to name this file `driver-class` also not sure if you need `class` you should be able to tell this is an object class?

File: index.js (hunk 2 above)
Comment (podviaznikov) [inline index.js:4]: Unfortunately I need to expose it here in order to stub functions on the base class.
I can't stub functions on the Docker or Swarm instance. I use this `_BaseClient` in other places where library is used

File: .travis.yml (hunk 1 above)
Comment (rsandor) [inline .travis.yml:4]: I think you can use `lts` here. @bkendall  would know more...

File: lib/logger.js (hunk 5 above)
Comment (rsandor) [inline lib/logger.js:27]: I thought we only use `LOG_LEVEL` now.

File: .travis.yml (hunk 1 above)
Comment (thejsj) [inline .travis.yml:4]: Also, if there's a way to test the compiled version agains v0.10, I'd do it.

Comment (podviaznikov) [inline .travis.yml:4]: I guess we can do it later if/when loki will try to make it way into our 0.10 code. I think we can start using loki in simpler services first that use Node 4

File: lib/base.js (hunk 4 above)
Comment (podviaznikov) [inline lib/base.js]: added Joi

File: .travis.yml (hunk 1 above)
Comment (anandkumarpatel) [inline .travis.yml:4]: +1 lets not support .10 (we are trying move away from it :) )

File: index.js (hunk 2 above)
Comment (anandkumarpatel) [inline index.js:4]: you sure? node just returns cached `require('./lib/base')` every time

File: lib/base.js [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+  // DOCKER_CERT_PATH is docker's default thing it checks - may as well use it
+  const certPath = process.env.DOCKER_CERT_PATH || '/etc/ssl/docker'
+  certs.ca = fs.readFileSync(join(certPath, '/ca.pem'))
+  certs.cert = fs.readFileSync(join(certPath, '/cert.pem'))
+  certs.key = fs.readFileSync(join(certPath, '/key.pem'))
+} catch (e) {
+  throw e
+}
+
+const opstSchema = joi.object({
```
Comment (anandkumarpatel) [inline lib/base.js]: `opstSchema` => `opsSchema`

File: lib/base.js (hunk 4 above)
Comment (anandkumarpatel) [inline lib/base.js]: type check DriverClass

**Comment by anandkumarpatel** [comment]
one nit, rest looks good to me

File: lib/base.js (hunk 7 above)
Comment (podviaznikov) [inline lib/base.js]: done

**Pull request merged**